### ISSUE TITLE
[BUG] CampanionPi sends multiple copies of string when using tcp send command

### ISSUE THREAD

**Issue by BaseballJim** [bitfocus/companion#1209]
Tried using the same button file on two instances of companion.  One on windows 10, the other CompanionPi. Latest builds as of today. Monitored the strings sent via tcp send command to target computer.  Win 10 sent 1 copy of each send command as expected.  CompanionPi sent two copies of each string. Duplicates are jumbled in time, sometimes one after another, sometimes intertwined with other strings.

**Expected behavior**

**Desktop (please complete the following information):**
 - OS: CompanionPi

**Comment by josephdadams** [comment]
@jarodwsams your thoughts here? This doesn't sound like it could necessarily be a CompanionPi problem.

@BaseballJim give a little more info on how you have the button set up. Generic TCP instance? Did you make the button fresh or did you import a config?

**Comment by jimamen** [comment]
@jarodwsams
Background:  The goal is to remotely control 3Play 425 which has NO built in remote mechanism at all.  It is able to lock the positions on its screen of buttons like 'play' 'stop' etc.   I installed a cop of MCEcontrol on 3Play that listens for commands on a TCP port and then sends mouse movements, clicks, keystrokes, etc to 3Play's operating system just like the mouse and keyboard connected to it.   I created the button page on a windows computer with buttons to select camera angle & play.  I am using Generic TCP to the IP of 3Play and port (I think its 5150) that MCEcontrol is listening on.   System works great.  Then I decided I don't want to haul a windows computer around with me (and I didn't have any laptops lying around) so I thought to put companion on a PI4 (which I do have several of, lying around).  I created the CompanionOS memory card and got the PI 4 to boot etc, etc  Then I loaded the saved configuration from the windows computer onto the PI.  Weird stuff happens, mouse moves to wrong place wrong buttons selected etc.  Spent a day trying to resolve that when I finally backed 3Play out to its OS (win7) and looked at MCEcontrol's log.   The log showed one string received for each string sent with the WINDOWS computer.  With Companion PI I get TWO strings received for each one sent.  Each one is duplicated but they aren't always received in the correct sequence.  The duplicated of string one may arrive after string 3 sometimes.    I tried creating a button from scratch on the PI with just a single string and it was as expected always received twice.  I have each string terminated with the Windows CR/LF since I assume its the receiver (3Play in my case) that cares about that and not the sender (Pi).

**Comment by josephdadams** [comment]
@BaseballJim so did you export the config from the Windows machine and import it on the CompanionPi?

**Comment by jimamen** [comment]
Yes, exactly.  Export from Windows, import to CompanionPI.   But I created a new button on the companionPi all by itself and it behaves the same way.       Without a GUI on CompanionPi I've been running it remotely from the same Windows computer  as  IP(CompanionPi):8000.    I use Chrome for a browser.      Technically I guess you could say I exported from the tab on Chrome pointing at http://127.0.0.1:8000/   and then imported from the same browser on Windows but with the tab pointing at IP(CompanionPi):8000.   Let me be clear lest the import/export thing dominate the thinking.         I created a single button on a CLEAN instance of CompanionPi with a SINGLE command and it was received as a duplicate each time.

**Comment by jimamen** [comment]
Let me be clear lest the import/export thing dominate the thinking. I created a single button on a CLEAN instance of CompanionPi with a SINGLE command (i.e. string)  and it was received as a duplicate each time (two strings both the same).   I also created a single button on a CLEAN instance of Companion 2.1.1 on Windows and those button presses were received correctly (i.e. a single string)

**Comment by josephdadams** [comment]
Have you tried to press the button through the emulator, tablet view, or “test actions” button to rule out a stream deck issue?

**Comment by jimamen** [comment]
I've used both the emulator and Streamdeck.  Its the same results regardless.  Pretty sure its not Streamdeck hardware.  Today I used the emulator exclusively.    I'm tempted to dig into the junk pile and find an Ethernet hub and then run wireshark on the line but I don't know what that would tell me.  I'm convinced duplicates are being sent.

**Comment by jimamen** [comment]
I thought  perhaps that CompanionPi had gotten two instances of "Generic TCP" instantiated but I checked and thats not the case (at least it only shows one on the "instance" page).    Is it possible that Companion is being started twice on the PI?   I didn't do anything out of the ordinary, just followed the steps on the web site.  I'm not sure two instances of the same program could bind to the same port to send TCP on anyway.

**Comment by josephdadams** [comment]
If two Companions were running you would be getting a lot of errors on port bindings.

**Comment by jarodwsams** [comment]
@jimamen have you run `sudo companion-update` since getting up and running? Possible there was an undocumented bug when I created that version of the image.

**Comment by jimamen** [comment]
Yes, like any good linux install it was the first thing I did after I got the pi to boot.

**Comment by jimamen** [comment]
I set up a separate Windows computer with the MCEcontrol app on it and installed Wireshark on it as well.  I created a network that only contained the Companion Pi  and this Windows computer.   CompanionPI ip=10.0.4.23, Windows computer 10.0.4.21.  The MCEcontrol app listens on port 5150 so I created a filter in Wireshark to capture only those ports.  I created a new "from scratch" button in CompanionPi that sends a single line to the target (Windows) computer on port 5150.   Wireshark shows TWO tcp strings for each (i.e. ONE) press of Companion's button.  See Wireshark screen capture attached.
![Wireshark capture](https://user-images.githubusercontent.com/68242594/91472113-a3be1480-e85c-11ea-809d-f3eabee3d25e.PNG)

**Comment by josephdadams** [comment]
What version of Companion are you running on each machine?

**Comment by jimamen** [comment]
Windows is: 2.1.1-2319482-2494PI: 2.1.1 (5c01252-2529) Let me just reiterate that the Wireshark screen capture above was created with a completely CLEAN Companion build of one button with one string on the Pi4 and had no import/export or any other interaction with the Windows machine (other than a Windows machine’s web browser was pointed to the IP of the PI4 since there is no GUI on the PI4).  I am convinced that the problem is ONLY on the PI4, not related to import/export or any other factor. Sent from Mail for Windows 10 From: Joseph AdamsSent: Tuesday, September 1, 2020 8:45 AMTo: bitfocus/companionCc: jimamen; MentionSubject: Re: [bitfocus/companion] [BUG] CampanionPi sends multiple copies of string when using tcp send command (#1209) What version of Companion are you running on each machine?—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by josephdadams** [comment]
We are tracking a double press issue in one of the latest releases, which is why I asked. Thanks.

**Comment by josephdadams** [comment]
Yes, I’m pretty sure your issue is caused by a recent update within Companion. If you upgraded your Windows install to latest, you would see it there as well. So you can roll back the CompanionPi install or wait for the bug fix. Any action is being sent twice in the latest release.

**Comment by jimamen** [comment]
Ok, I'll try rolling back the pi version.  Thanks for working the issue.
I'll probably not mess with the windows version, I've got a system that
works and think I'll keep it that way.

**Comment by josephdadams** [comment]
This is fixed in the latest release.

**Comment by jimamen** [comment]
FYI I reflashed the 2 July 2020 version but did NOT run the update command and it seems to work.
